NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1071,3 +1139,145 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length) const
+            sizeof(CExtendedNBStorage),
+            PARANDIS_MEMORY_TAG,
+            NormalPoolPriority);
+
+        if (m_ExtraNBStorage == nullptr)
+        {
+            DPrintf(0, "[%s] ExtendedNBStorage allocation failed \n", __FUNCTION__);
+            return false;
+        }
+        if (!BorrowPages(uPages))
+        {
+            return false;
+        }
+   
+        ULONG uDataOffset = 0;
+        for (ULONG i = 0; i < uPages; i++)
+        {
+            PVOID pVirtualAddress = m_ExtraNBStorage->m_UsedPages[i]->GetVA();
+            ULONG uCopySize = (i == uPages - 1) ? (uDataLength - uDataOffset) : PAGE_SIZE;
+            ULONG uCopied = Copy(pVirtualAddress, uCopySize, uDataOffset);
+            if (uCopied != uCopySize)
+            {
+                ReturnPages();
+                DPrintf(0, "[%s] copy failed! expected %lu, copied %lu bytes\n", __FUNCTION__, uCopySize, uCopied);
+                return false;
+            }
+            m_ExtraNBStorage->m_Elements[i].Address = m_ExtraNBStorage->m_UsedPages[i]->GetPA();
+            m_ExtraNBStorage->m_Elements[i].Length = uCopySize;
+            uDataOffset += uCopySize;  // Adjust offset, last page might not need adjustment but does not harm
+        }
+        return true;
+    }
+
+    bool CNB::MapCopyDataToVirtioSGL(CTXDescriptor &Descriptor, ULONG Offset) const
+    {
+        for (ULONG i = 0; i < m_ExtraNBStorage->m_UsedPagesCount; i++)
+        {
+            if (Offset < m_ExtraNBStorage->m_Elements[i].Length)
+            {
+                PHYSICAL_ADDRESS PA;
+                PA.QuadPart = m_ExtraNBStorage->m_Elements[i].Address.QuadPart + Offset;
+
+                if (!Descriptor.AddDataChunk(PA, m_ExtraNBStorage->m_Elements[i].Length - Offset))
+                {
+                    return false;
+                }
+                Offset = 0;
+            }
+            else
+            {
+                Offset -= m_ExtraNBStorage->m_Elements[i].Length;
+            }
+        }
+
+        return true;
+    }
+
+    bool CNB::BorrowPages(ULONG NeedPages)
+    {
+        CParaNdisTX * ParentTxPath = m_ParentNBL->GetParentTXPath();
+        if (ParentTxPath->GetAvailPagesCount() < NeedPages)
+        {
+            m_NBStatus = NBStatus::NB_NO_RESOURCE;
+            return false;
+        }
+        for (ULONG i = 0; i < NeedPages; i++)
+        {
+            CNdisSharedMemory* pPage = ParentTxPath->PopPage();
+            m_ExtraNBStorage->m_UsedPages[i] = pPage;
+        }
+        m_ExtraNBStorage->m_UsedPagesCount = NeedPages;
+        return true;
+    }
+
+    void CNB::ReturnPages() 
+    {
+        if (m_ExtraNBStorage)
+        {
+            for (ULONG i = 0; i < m_ExtraNBStorage->m_UsedPagesCount; i++)
+            {
+                m_ParentNBL->GetParentTXPath()->PushPage(m_ExtraNBStorage->m_UsedPages[i]);
+            }
+            m_ExtraNBStorage->m_UsedPagesCount = 0;
```

Comment:
BTW, we can free the entire m_ExtraNBStorage here, correct?

Reply:
indeed